File: Code/WebSvc.Java/src/main/java/services/ServiceBuilder.java
package services;

import repositories.IDb;
import repositories.InMemoryDb;
import repositories.MongoAdapter;

/**
 * Implementation for Local and PROD env (Factory pattern)
 */
public class ServiceBuilder implements ISvcBuilder {
	
	private static ServiceBuilder _instance = null;
	
	// Runtime env
	private IEnvironment _env = TheEnvironment.GetInstance();
	
	private ServiceBuilder() {
	}
	
	public static ISvcBuilder GetInstance() {
		if(_instance == null) _instance = new ServiceBuilder();		
		return _instance;
	}

	@Override
	public IDb createDb() {
		if(_env.IsLocal())
			return new InMemoryDb();
		return new MongoAdapter();
	}

	@Override
	public IPersonaService createPersonaService() {
		// We don't need to mock this in Local env. Always return the real one.
		return new PersonaService(createDb());
	}

}

File: Code/WebSvc/WebSvc/Services/ServiceBuilder.cs
using Repositories;

namespace Services
{
    public class ServiceBuilder : ISvcBuilder
    {
        private static ServiceBuilder _instance;
        private IEnvironment _env;

        public ServiceBuilder()
        {
            // TODO we will see how this is handled by a MW...
            this._env = new TheEnvironment();
        }

        public static ISvcBuilder GetInstance()
        {
            if (_instance == null)
                _instance = new ServiceBuilder();

            return _instance;
        }

        public IDb CreateDb()
        {
            if (_env.IsLocal())
                return new InMemoryDb();
            return new MongoDB();
        }
    }
}
